dbms/src/Storages/DeltaMerge/DeltaMergeStore_Ingest.cpp

```diff
@@ -210,6 +140,401 @@ void DeltaMergeStore::ingestFiles(
+
+        while (!remaining_delete_range.none())
+        {
+            SegmentPtr segment;
+            {
+                std::shared_lock lock(read_write_mutex);
+
+                auto segment_it = segments.upper_bound(remaining_delete_range.getStart());
+                RUNTIME_CHECK(segment_it != segments.end(), remaining_delete_range.toDebugString());
+                segment = segment_it->second;
+            }
+
+            const auto delete_range = remaining_delete_range.shrink(segment->getRowKeyRange());
+            RUNTIME_CHECK(
+                !delete_range.none(), // as remaining_delete_range is not none, we expect the shrinked range to be not none.
+                delete_range.toDebugString(),
+                segment->simpleInfo(),
+                remaining_delete_range.toDebugString());
+            LOG_FMT_DEBUG(
+                log,
+                "Table ingest using split - delete range phase - Try to delete range in segment, delete_range={} segment={} remaining_delete_range={} updated_segments_n={}",
+                delete_range.toDebugString(),
+                segment->simpleInfo(),
+                remaining_delete_range.toDebugString(),
+                updated_segments.size());
+
+            const bool succeeded = segment->write(*dm_context, delete_range);
+            if (succeeded)
+            {
+                updated_segments.insert(segment);
+                RUNTIME_CHECK(compare(delete_range.getEnd(), remaining_delete_range.getStart()) >= 0);
+                remaining_delete_range.setStart(delete_range.end); // We always move forward
+            }
+            else
+            {
+                // segment may be abandoned, retry current range by finding the segment again.
+            }
+        }
+    }
+
+    LOG_FMT_DEBUG(
+        log,
+        "Table ingest using split - delete range phase - finished, updated_segments_n={}",
+        updated_segments.size());
+
+    /*
+     * In second phase (SplitIngest Phase),
+     * we will try to ingest DMFile one by one into the segments in order.
+     *
+     * Consider the following case:
+     * -Inf                                                                         +Inf
+     *  │        │--------------- Ingest Range ---------------│                       │
+     *  │               │-- DMFile --│- DMFile --│---- DMFile ----│- DMFile --│       │
+     *  │- Segment --│-- Seg --│------- Segment -----│- Seg -│------- Segment --------│
+     *
+     * This is what we will ingest:
+     * Iterate 0:
+     * -Inf                                                                         +Inf
+     *  │        │--------------- Ingest Range ---------------│                       │
+     *  │               │-- DMFile --│                                                │
+     *  │            │-- Seg --│------- Segment -----│                                │
+     *                   ↑              ↑ The segment we ingest DMFile into
```

Comment:
```suggestion
     *                   ↑                The segment we ingest DMFile into
```

Reply:
Oh, may be I should made comments more clear. I would like to illustrate what we will iterate for the outmost loop `for (size_t file_idx = 0; file_idx < files.size(); file_idx++)`.